File: server/src/services/intervention-mapper.ts
// Intervention mapping service (ported from process_data.py lines 424-534)

const DQ_DELAY_THRESHOLD = 75;
const DQ_CONCERN_THRESHOLD = 85;

export interface InterventionActivity {
  domain: string;
  activityName: string;
  frequency: string;
  durationMinutes: number;
  caregiverFormat: string;
  priority: number;
  rationale: string;
}

export interface InterventionInput {
  lcDq: number | null;
  gmDq: number | null;
  fmDq: number | null;
  cogDq: number | null;
  seDq: number | null;
  behaviourRiskLevel: string | null;
  behaviourConcerns: string | null;
  behaviourScore: number;
  nutritionRisk: string | null;
  nutritionScore: number;
}

export function mapInterventions(input: InterventionInput): InterventionActivity[] {
  const plans: InterventionActivity[] = [];

  // Speech/Language intervention
  if (input.lcDq !== null && input.lcDq < DQ_CONCERN_THRESHOLD) {
    const intensity = input.lcDq < DQ_DELAY_THRESHOLD ? 'intensive' : 'moderate';
    plans.push({
      domain: 'Speech & Language',
      activityName: 'Structured speech stimulation with picture cards and story narration',
      frequency: intensity === 'intensive' ? 'Daily' : '3x/week',
      durationMinutes: intensity === 'intensive' ? 15 : 10,
      caregiverFormat: 'audio',
      priority: intensity === 'intensive' ? 1 : 2,
      rationale: `LC DQ=${Math.round(input.lcDq)} (below ${DQ_CONCERN_THRESHOLD})`,
    });
  }

  // Gross Motor intervention
  if (input.gmDq !== null && input.gmDq < DQ_CONCERN_THRESHOLD) {
    const intensity = input.gmDq < DQ_DELAY_THRESHOLD ? 'intensive' : 'moderate';
    plans.push({
      domain: 'Gross Motor',
      activityName: 'Structured physical movement exercises — crawling, climbing, balancing',
      frequency: intensity === 'intensive' ? 'Daily' : '3x/week',
      durationMinutes: intensity === 'intensive' ? 20 : 15,
      caregiverFormat: 'visual',
      priority: intensity === 'intensive' ? 1 : 2,
      rationale: `GM DQ=${Math.round(input.gmDq)} (below ${DQ_CONCERN_THRESHOLD})`,
    });
  }

  // Fine Motor intervention
  if (input.fmDq !== null && input.fmDq < DQ_CONCERN_THRESHOLD) {
    const intensity = input.fmDq < DQ_DELAY_THRESHOLD ? 'intensive' : 'moderate';
    plans.push({
      domain: 'Fine Motor',
      activityName: 'Bead threading, clay molding, crayon coloring, and buttoning exercises',
      frequency: intensity === 'intensive' ? 'Daily' : '4x/week',
      durationMinutes: 15,
      caregiverFormat: 'visual',
      priority: intensity === 'intensive' ? 1 : 2,
      rationale: `FM DQ=${Math.round(input.fmDq)} (below ${DQ_CONCERN_THRESHOLD})`,
    });
  }

  // Cognitive intervention
  if (input.cogDq !== null && input.cogDq < DQ_CONCERN_THRESHOLD) {
    const intensity = input.cogDq < DQ_DELAY_THRESHOLD ? 'intensive' : 'moderate';
    plans.push({
      domain: 'Cognitive',
      activityName: 'Pattern recognition games, shape sorting, problem-solving puzzles',
      frequency: intensity === 'intensive' ? 'Daily' : '3x/week',
      durationMinutes: intensity === 'intensive' ? 15 : 10,
      caregiverFormat: 'visual',
      priority: intensity === 'intensive' ? 1 : 2,
      rationale: `COG DQ=${Math.round(input.cogDq)} (below ${DQ_CONCERN_THRESHOLD})`,
    });
  }

  // Socio-emotional intervention
  if (input.seDq !== null && input.seDq < DQ_CONCERN_THRESHOLD) {
    const intensity = input.seDq < DQ_DELAY_THRESHOLD ? 'intensive' : 'moderate';
    plans.push({
      domain: 'Socio-Emotional',
      activityName: 'Group play activities, emotion-naming games, turn-taking exercises',
      frequency: intensity === 'intensive' ? '3x/week' : '2x/week',
      durationMinutes: 15,
      caregiverFormat: 'audio',
      priority: intensity === 'intensive' ? 1 : 3,
      rationale: `SE DQ=${Math.round(input.seDq)} (below ${DQ_CONCERN_THRESHOLD})`,
    });
  }

  // Behavioral intervention
  if (input.behaviourRiskLevel === 'High' || input.behaviourRiskLevel === 'Moderate') {
    let concern = input.behaviourConcerns ?? 'General behavioral regulation';
    if (concern === 'Unknown' || concern === 'None' || concern === '') {
      concern = 'General behavioral regulation';
    }
    plans.push({
      domain: 'Behavioral',
      activityName: `Targeted behavioral intervention for ${concern.toLowerCase()} — positive reinforcement and caregiver guidance`,
      frequency: input.behaviourRiskLevel === 'High' ? 'Daily' : '3x/week',
      durationMinutes: 10,
      caregiverFormat: 'audio',
      priority: 2,
      rationale: `Behaviour score=${input.behaviourScore}, risk=${input.behaviourRiskLevel}`,
    });
  }

  // Nutrition intervention
  if (input.nutritionRisk === 'High' || input.nutritionRisk === 'Medium') {
    plans.push({
      domain: 'Nutrition',
      activityName: 'Supplementary feeding program + growth monitoring + caregiver nutrition counseling',
      frequency: input.nutritionRisk === 'High' ? 'Daily' : '3x/week',
      durationMinutes: 0, // ongoing
      caregiverFormat: 'visual',
      priority: input.nutritionRisk === 'High' ? 1 : 2,
      rationale: `Nutrition score=${input.nutritionScore}, risk=${input.nutritionRisk}`,
    });
  }

  // Sort by priority
  plans.sort((a, b) => a.priority - b.priority);

  // Cap at 5 activities per child
  return plans.slice(0, 5);
}

File: risk-engine/app/services/intervention_mapper.py
"""
ECD Risk Engine — Intervention Recommendation Mapper
Ported from scripts/process_data.py lines 424-534.
"""

from app.models.schemas import ChildAssessmentInput, InterventionPlanOutput
from app.utils.constants import (
    DQ_DELAY_THRESHOLD,
    DQ_CONCERN_THRESHOLD,
    MAX_INTERVENTION_ACTIVITIES,
)


def generate_intervention_plan(child: ChildAssessmentInput) -> list[InterventionPlanOutput]:
    """
    Generate intervention plans for a child based on DQ scores and risk levels.
    Exact same logic as process_data.py lines 429-534.
    Capped at MAX_INTERVENTION_ACTIVITIES, sorted by priority.
    """
    plans: list[InterventionPlanOutput] = []

    # Speech/Language intervention
    if child.lc_dq < DQ_CONCERN_THRESHOLD and child.lc_dq > 0:
        severity = "intensive" if child.lc_dq < DQ_DELAY_THRESHOLD else "moderate"
        plans.append(InterventionPlanOutput(
            domain="Speech & Language",
            activity="Structured speech stimulation with picture cards and story narration",
            frequency="Daily" if severity == "intensive" else "3x/week",
            duration_minutes=15 if severity == "intensive" else 10,
            caregiver_format="audio",
            priority=1 if severity == "intensive" else 2,
            rationale=f"LC DQ={child.lc_dq:.0f} (below {DQ_CONCERN_THRESHOLD})",
        ))

    # Gross Motor intervention
    if child.gm_dq < DQ_CONCERN_THRESHOLD and child.gm_dq > 0:
        severity = "intensive" if child.gm_dq < DQ_DELAY_THRESHOLD else "moderate"
        plans.append(InterventionPlanOutput(
            domain="Gross Motor",
            activity="Structured physical movement exercises — crawling, climbing, balancing",
            frequency="Daily" if severity == "intensive" else "3x/week",
            duration_minutes=20 if severity == "intensive" else 15,
            caregiver_format="visual",
            priority=1 if severity == "intensive" else 2,
            rationale=f"GM DQ={child.gm_dq:.0f} (below {DQ_CONCERN_THRESHOLD})",
        ))

    # Fine Motor intervention
    if child.fm_dq < DQ_CONCERN_THRESHOLD and child.fm_dq > 0:
        severity = "intensive" if child.fm_dq < DQ_DELAY_THRESHOLD else "moderate"
        plans.append(InterventionPlanOutput(
            domain="Fine Motor",
            activity="Bead threading, clay molding, crayon coloring, and buttoning exercises",
            frequency="Daily" if severity == "intensive" else "4x/week",
            duration_minutes=15,
            caregiver_format="visual",
            priority=1 if severity == "intensive" else 2,
            rationale=f"FM DQ={child.fm_dq:.0f} (below {DQ_CONCERN_THRESHOLD})",
        ))

    # Cognitive intervention
    if child.cog_dq < DQ_CONCERN_THRESHOLD and child.cog_dq > 0:
        severity = "intensive" if child.cog_dq < DQ_DELAY_THRESHOLD else "moderate"
        plans.append(InterventionPlanOutput(
            domain="Cognitive",
            activity="Pattern recognition games, shape sorting, problem-solving puzzles",
            frequency="Daily" if severity == "intensive" else "3x/week",
            duration_minutes=15 if severity == "intensive" else 10,
            caregiver_format="visual",
            priority=1 if severity == "intensive" else 2,
            rationale=f"COG DQ={child.cog_dq:.0f} (below {DQ_CONCERN_THRESHOLD})",
        ))

    # Socio-Emotional intervention
    if child.se_dq < DQ_CONCERN_THRESHOLD and child.se_dq > 0:
        severity = "intensive" if child.se_dq < DQ_DELAY_THRESHOLD else "moderate"
        plans.append(InterventionPlanOutput(
            domain="Socio-Emotional",
            activity="Group play activities, emotion-naming games, turn-taking exercises",
            frequency="3x/week" if severity == "intensive" else "2x/week",
            duration_minutes=15,
            caregiver_format="audio",
            priority=1 if severity == "intensive" else 3,
            rationale=f"SE DQ={child.se_dq:.0f} (below {DQ_CONCERN_THRESHOLD})",
        ))

    # Behavioral intervention
    if child.behaviour_risk_level in ("High", "Moderate"):
        concern = child.behaviour_concerns
        if concern in ("Unknown", "None", ""):
            concern = "General behavioral regulation"
        plans.append(InterventionPlanOutput(
            domain="Behavioral",
            activity=f"Targeted behavioral intervention for {concern.lower()} — positive reinforcement and caregiver guidance",
            frequency="Daily" if child.behaviour_risk_level == "High" else "3x/week",
            duration_minutes=10,
            caregiver_format="audio",
            priority=2,
            rationale=f"Behaviour score={child.behaviour_score}, risk={child.behaviour_risk_level}",
        ))

    # Nutrition intervention
    if child.nutrition_risk in ("High", "Medium"):
        plans.append(InterventionPlanOutput(
            domain="Nutrition",
            activity="Supplementary feeding program + growth monitoring + caregiver nutrition counseling",
            frequency="Daily" if child.nutrition_risk == "High" else "3x/week",
            duration_minutes=0,  # ongoing
            caregiver_format="visual",
            priority=1 if child.nutrition_risk == "High" else 2,
            rationale=f"Nutrition score={child.nutrition_score}, risk={child.nutrition_risk}",
        ))

    # Sort by priority and cap at MAX_INTERVENTION_ACTIVITIES
    plans.sort(key=lambda x: x.priority)
    return plans[:MAX_INTERVENTION_ACTIVITIES]
